### PR TITLE
[CR] [WIP] Battery overhaul

### LINKED ISSUE #29536: Proposal for a battery overhaul
**Is your feature request related to a problem? Please describe.**  
Batteries in this game are a little abstract at the moment. I find that kind of annoying and would like to propose an overhaul of the battery system. The major problems I currently see are as follows:

* Everything from an mp3 player to an arc furnace accepts the same kind of batteries.
* You can split a flashlight battery 100 ways if you want to for some reason.

After doing some research, it turns out that tools that run off of batteries can be modded to accept magazines with no extra C++ code whatsoever. I've created this mod as a proof of concept:

[MagazineFlashlight.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/3076715/MagazineFlashlight.zip)

**Describe the solution you'd like**  
All battery powered tools should be converted to run off of magazines rather than raw batteries. My proposal involves three types:

- **Light personal batteries** - These can fit in the palm of your hand and are used to power small electronics like flashlights, mp3 players, and watches.
- **Medium personal batteries** - These are about the size of your fist and are used to power commercially available power tools and kitchen appliances. 
- **Large personal batteries** - These are about the size of car batteries and can be used to power industrial grade power tools like furnaces, kilns, and jackhammers. 

**Describe alternatives you've considered**  
The battery system works well enough right now, but it's kind of silly. Since it seems like all we would need is some JSON for the lion's share of this proposal to work, this might not be as hard as you'd think.

**Additional context**  
[MagazineFlashlight.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/3076715/MagazineFlashlight.zip)

Additionally, [this is a link for a comparison from my branch to the master branch.](https://github.com/CleverRaven/Cataclysm-DDA/compare/master...Inglonias:battery_overhaul)

**Risks and Challenges**
While code-wise, this may not be too challenging, this proposal has a number of problems I haven't actually worked out yet.
1. What do we do about existing save games? If we migrate over carelessly, the best that could happen is that people can load up their existing games and decide themselves how to use the debug menu to migrate to the new system. That's... not great.
2. Modded tools are a problem. The extra battery mod can be replaced with high capacity versions of these batteries (maybe just taping two light batteries together to make a high-capacity version), but what happens to UPS conversions and vehicle battery conversions?

### PR DESCRIPTION
**Pull request by Inglonias**
#### Summary
```SUMMARY: Balance "Adds a mod that replaces the existing battery system"```

#### Purpose of change
Fixes #29536 

#### Describe the solution
This PR adds a mod that replaces the somewhat abstract battery system with one that works similarly to how tools already work after they have a vehicle battery compartment mod. All tools now accept batteries as magazines. Different tools accept different sizes of batteries. Cell phones can no longer use the same kind of battery as an electric furnace or welder. There is at least one variant of each size of battery.

You can find batteries in existing tools as they spawn. It should work just fine there. Feel free to unload these tools as normal.

**Task List**
- [x] Add new battery types as items
- [x] Ensure all battery powered tools are compatible with new battery items
- [x] Ensure all tools are rebalanced properly to work with new batteries
- [x] Allow survivors to craft relevant adapters as tool mods
- [x] Allow survivors to assemble and disassemble batteries using recipes.
- [x] Place new batteries in existing item groups for spawning
- [x] Place new batteries in existing mapgen entries for spawning

#### Describe alternatives you've considered
The current battery system works, but could be vastly improved to make more sense and be more interesting from a gameplay standpoint.
#### Additional context
The purpose of adding this content as a mod is to postpone the inevitable headache that migrating existing saved games will cause, and to make it optional for those who want to wait until all the kinks get worked out. To be honest, I don't know if a migration is even possible.

The battery compatibility chart is below:
image

### CODE REVIEW

**Comment by Inglonias** [comment]
jenkins rebuild

**Comment by nexusmrsep** [comment]
Love the idea, but I'd rather see 3LR12 (4.5-volt), D, C, AA, AAA, AAAA, A23 (12-volt), PP3 (9-volt), CR2032 (3-volt), and LR44 batteries, instead of some generic names batteries that my mind cannot connect to real world items.
image

**Comment by Inglonias** [comment]
> Love the idea, but I'd rather see 3LR12 (4.5-volt), D, C, AA, AAA, AAAA, A23 (12-volt), PP3 (9-volt), CR2032 (3-volt), and LR44 batteries, instead of some generic names batteries that my mind cannot connect to real world items.

I'm not using IRL battery names because I'm not knowledgeable enough to pull that off.
That said, I'm not against your suggestion if you can provide me good names for each battery type.

**Comment by Inglonias** [comment]
jenkins rebuild

**Comment by Inglonias** [comment]
jenkins rebuild

**Comment by Inglonias** [comment]
I'd also like to avoid having ten different types of battery in the game. Ideally, I want to stick to the 3.5 I've already got if I can.

**Comment by Inglonias** [comment]
jenkins rebuild

**Comment by ifreund** [comment]
You probably want to lint all you json, not just one at a time.  Jenkins will bail on the first error it finds and not inform about subsequent errors.

**Comment by Inglonias** [comment]
I did. It doesn't seem to matter.

Someone,on the discord pointed out its an issue with the carriage returns in the json. Tried removing them and my git repo had a fit. I am re cloning now.

**Comment by dissociativity** [comment]
I'd go for Coin Cell, AA and 9v for non-rechargeable batteries if you want 3.
Arbitrary sizes make less sense to most people.
I'd like for in future for rechargeable cells to generally be made from 3.7v 18650 lithium ion cells, and that's how one would repair storage batteries, by disassembling another storage battery or other rechargeable battery and then making a new storage battery or consuming an 18650 and some solder/soldering iron charges.

**Comment by nexusmrsep** [comment]
I agree with @dissociativity, but I'd also add AAA batteries. They are commonly used in small devices like mp3 players.

There is however another problem with bringing up RL batteries, and it's the fact that awful lot of devices use multiple number of those, and that wouldn't be replicable under existing magazines handling code.

**Comment by Inglonias** [comment]
> I'd go for Coin Cell, AA and 9v for non-rechargeable batteries if you want 3.
> Arbitrary sizes make less sense to most people.

I agree with the latter point, but not the former. I'd slot that in as "ultralight", "light", and "medium" respectively in the current system if I absolutely had to. Unfortunately, I don't see a welder running off of a single 9v battery, and it's currently set up for medium batteries.

My main issue with using real battery types is that I don't see a real analogue we can use that would both make sense to people and that would be as simple as the system I've got. In real life, there's a million different batteries in all shapes and sizes, and reproducing that in the game would get really annoying really quickly, especially if you're used to the old system where everything uses the same kind of battery.

If you say an ultralight battery is such and such kind of coin cell, people ask "why can I use that in a cell phone?" The real answer is "because it makes for less frustrating gameplay. than not being able to" You can't slot a coin cell battery into a cell phone and expect it to work in real life, but in this implementation, you can. Nobody in their right mind would make a battery-powered kiln or electric furnace, but because of the way tools currently work, that's the best option we currently have in this game.

Let me explain how I see each battery size in my current implementation, and we can discuss real world names from there:

**Ultralight:** I see this like a modern coin cell battery. Used where size is important or the manufacturers are cheap bastards (which is why it's standard in the talking doll)
image. I would be willing to cleanly break this apart from light if that's what people want, but for now, light devices can fit these, but ultralight ones can't fit light batteries.

**Light:** I see this as a smartphone battery, basically. The high-capacity variant is a high-end one for consumer goods that need a bit more juice like professional camera equipment.
image

**Medium:** I see this like those power packs for cordless tools. The high capacity ones are for things like military-grade equipment (mess kits) and fancier electronics (L-sticks)
image

**Heavy:** These are just car batteries in all but name. The high capacity heavy batteries currently only come standard on a UPS
image

I'm sorry if it feels like I'm being stubborn here, but that's my reasoning. It breaks my immersion to be able to load a cell phone with a coin cell, or a power drill with a 9V battery.

**Comment by Zireael07** [comment]
> It breaks my immersion to be able to load a cell phone with a coin cell, or a power drill with a 9V battery.

Agreed, IMHO the generic names are better. It's the future anyway and we can't tell what batteries will be in use by then.

File: data/mods/MagazineFlashlight/batt_ammo.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[
```
Comment (Inglonias) [inline data/mods/MagazineFlashlight/batt_ammo.json]: This whole directory was a proof of concept and should not be here.

**Comment by Night-Pryanik** [comment]
> The purpose of adding this content as a mod is to postpone the inevitable headache that migrating existing saved games will cause,

I believe our current battery system NEEDS to be changed. Your PR is a good candidate for this change. Obviously it is planned to be mainlined sometime in the future, and the sooner it happens, the better. You're right to say it's inevitable, so I don't see a see a reason to postpone it.

> to make it optional for those who want to wait until all the kinks get worked out.

For this it would be better to mainline your changes and make a mod that returns the old system.

**Comment by Inglonias** [comment]
> I believe our current battery system NEEDS to be changed. Your PR is a good candidate for this change. Obviously it is planned to be mainlined sometime in the future, and the sooner it happens, the better. You're right to say it's inevitable, so I don't see a see a reason to postpone it.

If that's the way you want to do things, I'll need to redo this PR accordingly, which I think means that I'm closing this for now. Frustrating, but definitely possible. The migration mod can be as simple as making it so new-style battery magazines are reloadable with old style batteries, so people can use up their existing stockpiles. I'll have to test how existing tools behave when they're replaced by the new system, too.

**Pull request closed without merging**